### PR TITLE
Transporter

### PR DESCRIPTION
**Pull request by shay124578**

### CODE REVIEW

File: src/main/java/frc/trigon/robot/subsystems/transporter/Transporter.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
+package frc.trigon.robot.subsystems.transporter;
+
+import com.ctre.phoenix.motorcontrol.ControlMode;
+import com.ctre.phoenix.motorcontrol.can.WPI_TalonSRX;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+
+public class Transporter extends SubsystemBase {
+
+
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/subsystems/transporter/Transporter.java:15]: extra newlines

File: src/main/java/frc/trigon/robot/subsystems/transporter/Transporter.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    private TransporterConstants.TransporterState currentState;
+    private final static Transporter INSTANCE = new Transporter();
+    private final WPI_TalonSRX motor;
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/subsystems/transporter/Transporter.java]: missing newline

File: src/main/java/frc/trigon/robot/subsystems/transporter/Transporter.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    public static Transporter getInstance() {
+        return INSTANCE;
+    }
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/subsystems/transporter/Transporter.java:18]: missing newline

File: src/main/java/frc/trigon/robot/subsystems/transporter/Transporter.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    private Transporter() {
+        motor = TransporterConstants.MOTOR;
+    }
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/subsystems/transporter/Transporter.java]: missing newline

File: src/main/java/frc/trigon/robot/subsystems/transporter/Transporter.java [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    public void setState (TransporterConstants.TransporterState state){
+        motor.set(ControlMode.PercentOutput,state.voltage/TransporterConstants.VOLTAGE_COMPENSATION);
+        currentState = state;
+    }
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/subsystems/transporter/Transporter.java]: missing newline

File: src/main/java/frc/trigon/robot/subsystems/transporter/Transporter.java [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+    public TransporterConstants.TransporterState getState(){
+        return currentState;
+    }
+
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/subsystems/transporter/Transporter.java:26]: extra newlines

File: src/main/java/frc/trigon/robot/subsystems/transporter/TransporterConstants.java [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+        LOAD_VOLTAGE = 300,
+        EJECT_VOLTAGE = -300;
+    private static final boolean MOTOR_INVERTED = false;
+    protected static final WPI_TalonSRX MOTOR = new WPI_TalonSRX(MOTOR_ID);
+    protected static final double VOLTAGE_COMPENSATION = 10.5;
+    static {
+        MOTOR.setInverted(MOTOR_INVERTED);
+        MOTOR.enableVoltageCompensation(true);
+        MOTOR.configVoltageCompSaturation(VOLTAGE_COMPENSATION);
+    }
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/subsystems/transporter/TransporterConstants.java:15]: missing newline

File: src/main/java/frc/trigon/robot/subsystems/transporter/TransporterConstants.java [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+import com.ctre.phoenix.motorcontrol.can.WPI_TalonSRX;
+
+public class TransporterConstants {
+    private static final int MOTOR_ID = 0;
+    private static final double
+        LOAD_VOLTAGE = 300,
+        EJECT_VOLTAGE = -300;
+    private static final boolean MOTOR_INVERTED = false;
+    protected static final WPI_TalonSRX MOTOR = new WPI_TalonSRX(MOTOR_ID);
+    protected static final double VOLTAGE_COMPENSATION = 10.5;
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/subsystems/transporter/TransporterConstants.java]: missing newline

Comment (laviRZ) [inline src/main/java/frc/trigon/robot/subsystems/transporter/TransporterConstants.java]: drop the protected

File: src/main/java/frc/trigon/robot/subsystems/transporter/TransporterConstants.java [hunk 9]
```diff
@@ -0,0 +1,29 @@
+package frc.trigon.robot.subsystems.transporter;
+
+import com.ctre.phoenix.motorcontrol.can.WPI_TalonSRX;
+
+public class TransporterConstants {
+    private static final int MOTOR_ID = 0;
+    private static final double
+        LOAD_VOLTAGE = 300,
+        EJECT_VOLTAGE = -300;
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/subsystems/transporter/TransporterConstants.java]: voltage should be between `VOLTAGE_COMPENSATION` AND `-VOLTAGE_COMPENSATION`

File: src/main/java/frc/trigon/robot/subsystems/transporter/TransporterConstants.java [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+    static {
+        MOTOR.setInverted(MOTOR_INVERTED);
+        MOTOR.enableVoltageCompensation(true);
+        MOTOR.configVoltageCompSaturation(VOLTAGE_COMPENSATION);
+    }
+    public enum TransporterState{
+        LOAD(LOAD_VOLTAGE),
+        EJECT(EJECT_VOLTAGE),
+        OFF(0);
+
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/subsystems/transporter/TransporterConstants.java:16]: extra newline

File: src/main/java/frc/trigon/robot/subsystems/transporter/TransporterConstants.java [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+
+        final double voltage;
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/subsystems/transporter/TransporterConstants.java]: missing newline

File: src/main/java/frc/trigon/robot/subsystems/transporter/TransporterConstants.java [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+
+import com.ctre.phoenix.motorcontrol.can.WPI_TalonSRX;
+
+public class TransporterConstants {
+    private static final int MOTOR_ID = 0;
+    private static final double
+        LOAD_VOLTAGE = 300,
+        EJECT_VOLTAGE = -300;
+    private static final boolean MOTOR_INVERTED = false;
+    protected static final WPI_TalonSRX MOTOR = new WPI_TalonSRX(MOTOR_ID);
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/subsystems/transporter/TransporterConstants.java]: should be last

**Review by laviRZ — CHANGES_REQUESTED** [review]
Format the code before committing

File: src/main/java/frc/trigon/robot/subsystems/transporter/Transporter.java [hunk 13]
```diff
@@ -0,0 +1,28 @@
+package frc.trigon.robot.subsystems.transporter;
+
+import com.ctre.phoenix.motorcontrol.ControlMode;
+import com.ctre.phoenix.motorcontrol.can.WPI_TalonSRX;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+
+public class Transporter extends SubsystemBase {
+    private TransporterConstants.TransporterState currentState;
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/subsystems/transporter/Transporter.java]: Move to the bottom of the variables

File: src/main/java/frc/trigon/robot/subsystems/transporter/Transporter.java [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+    private final static Transporter INSTANCE = new Transporter();
+    private final WPI_TalonSRX motor;
+
+    public static Transporter getInstance() {
+        return INSTANCE;
+    }
+
+    private Transporter() {
+        motor = TransporterConstants.MOTOR;
+    }
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/subsystems/transporter/Transporter.java]: should be the top function

File: src/main/java/frc/trigon/robot/subsystems/transporter/Transporter.java [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+
+    public void setState(TransporterConstants.TransporterState state) {
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/subsystems/transporter/Transporter.java]: static import

File: src/main/java/frc/trigon/robot/subsystems/transporter/Transporter.java [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+        motor.set(ControlMode.PercentOutput, state.voltage / TransporterConstants.VOLTAGE_COMPENSATION);
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/subsystems/transporter/Transporter.java]: make a function in utilities.Conversions called `voltageToCompensatedPower` to do this calculation

File: src/main/java/frc/trigon/robot/subsystems/transporter/Transporter.java [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+        currentState = state;
+    }
+    public TransporterConstants.TransporterState getState() {
+        return currentState;
+    }
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/subsystems/transporter/Transporter.java]: get before set

File: src/main/java/frc/trigon/robot/subsystems/transporter/TransporterConstants.java [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+import com.ctre.phoenix.motorcontrol.can.WPI_TalonSRX;
+
+public class TransporterConstants {
+    private static final int MOTOR_ID = 0;
+    private static final double
+            LOAD_VOLTAGE = 10,
+            EJECT_VOLTAGE = -10;
+    private static final boolean MOTOR_INVERTED = false;
+    static final double VOLTAGE_COMPENSATION = 10.5;
+    protected static final WPI_TalonSRX MOTOR = new WPI_TalonSRX(MOTOR_ID);
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/subsystems/transporter/TransporterConstants.java]: default access, not protected

File: src/main/java/frc/trigon/robot/subsystems/transporter/TransporterConstants.java [hunk 19]
```diff
@@ -0,0 +1,31 @@
...
+    static {
+        MOTOR.setInverted(MOTOR_INVERTED);
+        MOTOR.enableVoltageCompensation(true);
+        MOTOR.configVoltageCompSaturation(VOLTAGE_COMPENSATION);
+    }
+
+    public enum TransporterState {
+        LOAD(LOAD_VOLTAGE),
+        EJECT(EJECT_VOLTAGE),
+        OFF(0);
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/subsystems/transporter/TransporterConstants.java:20]: missing newline

File: src/main/java/frc/trigon/robot/subsystems/transporter/TransporterConstants.java [hunk 20]
```diff
@@ -0,0 +1,31 @@
...
+
+import com.ctre.phoenix.motorcontrol.can.WPI_TalonSRX;
+
+public class TransporterConstants {
+    private static final int MOTOR_ID = 0;
+    private static final double
+            LOAD_VOLTAGE = 10,
+            EJECT_VOLTAGE = -10;
+    private static final boolean MOTOR_INVERTED = false;
+    static final double VOLTAGE_COMPENSATION = 10.5;
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/subsystems/transporter/TransporterConstants.java]: should be on top

**Review by laviRZ — CHANGES_REQUESTED** [review]
reformat your code before committing

File: src/main/java/frc/trigon/robot/subsystems/transporter/Transporter.java [hunk 21]
```diff
@@ -0,0 +1,34 @@
...
+        return INSTANCE;
+    }
+
+    public TransporterState getState() {
+        return currentState;
+    }
+
+    public void setState(TransporterState state) {
+        motor.set(
+                ControlMode.PercentOutput, Conversions.voltageToCompensatedPower(state.voltage, VOLTAGE_COMPENSATION));
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/subsystems/transporter/Transporter.java]: move to a separate line

File: src/main/java/frc/trigon/robot/subsystems/transporter/TransporterConstants.java [hunk 22]
```diff
@@ -0,0 +1,30 @@
+package frc.trigon.robot.subsystems.transporter;
+
+import com.ctre.phoenix.motorcontrol.can.WPI_TalonSRX;
+
+public class TransporterConstants {
+    static final double VOLTAGE_COMPENSATION = 10.5;
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/subsystems/transporter/TransporterConstants.java]: rename to voltage compensation saturation

File: src/main/java/frc/trigon/robot/utilities/Conversions.java [hunk 23]
```diff
@@ -0,0 +1,7 @@
+package frc.trigon.robot.utilities;
+
+public class Conversions {
+    public static double voltageToCompensatedPower(double currentVoltage, double voltageCompensation) {
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/utilities/Conversions.java]: rename to voltage compensation saturation

**Review by laviRZ — CHANGES_REQUESTED** [review]

File: src/main/java/frc/trigon/robot/subsystems/transporter/Transporter.java [hunk 24]
```diff
@@ -0,0 +1,34 @@
...
+import static frc.trigon.robot.subsystems.transporter.TransporterConstants.TransporterState;
+import static frc.trigon.robot.subsystems.transporter.TransporterConstants.VOLTAGE_COMPENSATION;
+
+public class Transporter extends SubsystemBase {
+    private final static Transporter INSTANCE = new Transporter();
+    private final WPI_TalonSRX motor;
+    private TransporterState currentState;
+
+    private Transporter() {
+        motor = TransporterConstants.MOTOR;
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/subsystems/transporter/Transporter.java]: move to decleration

File: src/main/java/frc/trigon/robot/subsystems/transporter/Transporter.java [hunk 25]
```diff
@@ -0,0 +1,38 @@
...
+
+import static frc.trigon.robot.subsystems.transporter.TransporterConstants.TransporterState;
+
+public class Transporter extends SubsystemBase {
+    private final static Transporter INSTANCE = new Transporter();
+    private final WPI_TalonSRX motor = TransporterConstants.MOTOR;
+    private TransporterState currentState;
+
+    private Transporter() {
+
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/subsystems/transporter/Transporter.java]: extra newline

File: src/main/java/frc/trigon/robot/subsystems/transporter/Transporter.java [hunk 26]
```diff
@@ -0,0 +1,38 @@
...
+    }
+
+    public static Transporter getInstance() {
+        return INSTANCE;
+    }
+
+    /**
+     * @return currentState
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/subsystems/transporter/Transporter.java]: unwritten doc

File: src/main/java/frc/trigon/robot/subsystems/transporter/Transporter.java [hunk 27]
```diff
@@ -0,0 +1,38 @@
...
+     */
+    public TransporterState getState() {
+        return currentState;
+    }
+
+    /**
+     * Set the motor power
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/subsystems/transporter/Transporter.java]: it sets the state, not the power

File: src/main/java/frc/trigon/robot/subsystems/transporter/Transporter.java [hunk 28]
```diff
@@ -0,0 +1,38 @@
...
+     *
+     * @param state TransporterState
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/subsystems/transporter/Transporter.java]: what?

**Review by laviRZ — CHANGES_REQUESTED** [review]

File: src/main/java/frc/trigon/robot/subsystems/transporter/Transporter.java (hunk 26 above)
Comment (shay124578) [reply src/main/java/frc/trigon/robot/subsystems/transporter/Transporter.java]: Yishai told me to write there nothing

Comment (laviRZ) [reply src/main/java/frc/trigon/robot/subsystems/transporter/Transporter.java]: Then I'm telling you to

File: src/main/java/frc/trigon/robot/subsystems/transporter/Transporter.java [hunk 29]
```diff
@@ -0,0 +1,37 @@
...
+
+public class Transporter extends SubsystemBase {
+    private final static Transporter INSTANCE = new Transporter();
+    private final WPI_TalonSRX motor = TransporterConstants.MOTOR;
+    private TransporterState currentState;
+
+    private Transporter() {
+    }
+
+    public static Transporter getInstance() {
```
Comment (lycaon7) [inline src/main/java/frc/trigon/robot/subsystems/transporter/Transporter.java:16]: Add Java docs

**Review by lycaon7 — DISMISSED** [review]

File: src/main/java/frc/trigon/robot/subsystems/transporter/Transporter.java [hunk 30]
```diff
@@ -0,0 +1,37 @@
...
+        return INSTANCE;
+    }
+
+    /**
+     * @return the current state of the transporterState
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/subsystems/transporter/Transporter.java]: the current state of the transporterState?

File: src/main/java/frc/trigon/robot/subsystems/transporter/Transporter.java [hunk 31]
```diff
@@ -0,0 +1,37 @@
...
+     */
+    public TransporterState getState() {
+        return currentState;
+    }
+
+    /**
+     * Set the motor state.
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/subsystems/transporter/Transporter.java]: doesn't set the motor state

File: src/main/java/frc/trigon/robot/subsystems/transporter/Transporter.java [hunk 32]
```diff
@@ -0,0 +1,37 @@
...
+     *
+     * @param state the wanted state for the transporter
+     */
+    public void setState(TransporterState state) {
+        motor.set(state.power);
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/subsystems/transporter/Transporter.java:33]: voltage compensation

File: src/main/java/frc/trigon/robot/subsystems/transporter/TransporterConstants.java [hunk 33]
```diff
@@ -0,0 +1,28 @@
...
+    static final double VOLTAGE_COMPENSATION_SATURATION = 10.5;
+    private static final int MOTOR_ID = 0;
+    static final WPI_TalonSRX MOTOR = new WPI_TalonSRX(MOTOR_ID);
+    private static final double
+            LOAD_POWER = 0.4,
+            EJECT_POWER = -0.4;
+    private static final boolean MOTOR_INVERTED = false;
+
+    static {
+        MOTOR.setInverted(MOTOR_INVERTED);
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/subsystems/transporter/TransporterConstants.java:14]: voltage compensation

**Review by laviRZ — CHANGES_REQUESTED** [review]

**Review by laviRZ — APPROVED** [review]

**Pull request merged**